UI/ChatGPT/README.md

```diff
@@ -0,0 +1,122 @@
+
+### How to create immutable records
+
+You can create immutable records in two ways. First, declare your record with a primary constructor and parameters; this will create an immutable record with the specified parameters as its properties:
+
+```csharp
+public partial record ChatResponse(string Message, bool IsError);
+```
+
+Another way is by creating properties using the `init` keyword instead of `set` to enforce immutability. Here's a brief example:
+
+```csharp
+public partial record ChatResponse
+{
+    public string Message { get; init; }
+    public bool IsError { get; init; }
+}
+```
+
+When you create a record, if you let the properties change with the `set` keyword, the record won't be immutable. This means you won't be able to lock in or keep the values from changing once you've set them.
+
+### How to use records with MVUX
+
+Records can be instantiated from the presentation layer as a parameter for a request or in the business layer, where data is usually retrieved/processed. For example, in our `ChatService`, we would have the following method being called from the Model. A `ChatEntry` list is received as a parameter from the Model, where `ChatEntry` is also a record, and we create an instance of `ChatResponse`, returning it to the Model:
+
+```csharp
+public async ValueTask<ChatResponse> AskAsync(IImmutableList<ChatEntry> history)
+{
+    var request = CreateRequest(history);
+
+    var result = await _client.CreateCompletion(request);
+
+    if (result.Successful)
+    {
+        var response = result.Choices.Select(choice => choice.Message.Content);
+
+        return new ChatResponse(string.Join("", responseContent));
+    }
+    else
+    {
+        return new ChatResponse(result.Error?.Message, IsError: true);
+    }
+}
+```
+
+### Updating records
+
+As we are dealing with immutable records, it's not possible to update them or their properties. To achieve that, we need to create a new instance based on the previous record. This ensures we are not modifying data from the UI in the wrong thread. See the example:
+
+Given the `Message` record:
+
+```csharp
+public partial record Message(string Content, Status status, Source source);
+```
+
+In our Model:
+
+```csharp
+var message = new Message("Hello world!", Status.Value, Source.User);
+
+...
+
+message = message with
+{
+    Content = response.Message,
+    Status = response.IsError ? Status.Error : Status.Value
+};
+
+//Then you can update your message list displayed in the UI, thread-safe
+await Messages.UpdateAsync(message);
+...
+
+```
+
+<!-- DIAGRAM SECTION -->
+## Architecture Diagram - WIP
+
+<picture>
+  <source media="(prefers-color-scheme: dark)" srcset="images/diagram-dark.png">
+  <source media="(prefers-color-scheme: light)" srcset="images/diagram-light.png">
+  <img alt="Sequence diagram" src="images/diagram-light.png">
+</picture>
```

Comment:
Why you're not using [GitHub's Mermaid extensions](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams)? It supports sequenceDiagram.

Examples: https://github.com/JakeSteam/Mermaid/blob/main/sequence.md

Reply:
Thanks Carl, it looks a lot better now. Changes pushed in next commit!